apex/contrib/csrc/groupbn/nhwc_batch_norm_kernel.h

```diff
@@ -426,69 +425,52 @@ DEVICE_FUNCTION void parallel_sums_16x2(float *smem, float (&x)[4], int nhw, voi
-#endif
+        // probably could do it earlier, before sync
+
+        for (int sync_iter=0; sync_iter < sync_iters; ++sync_iter) {
+            //float* params_pair_data = (reinterpret_cast<float**>(params_pair_datas))[sync_iter];
+            void* params_pair_data = params_pair_datas[sync_iter];
+
+            // skip the space consumed by previous sync iterations
+            const int xbuf_offset = sync_iter*data_total;
+            // data starts after flags, but have to skip previous
+            const int data_offset = xbuf_offset
+                                    + off*ELEMENTS_PER_LDG*THREADS_PER_PIXEL*2
+                                    + ELEMENTS_PER_LDG*threadIdx.x*2;
+
+            // after sums for this GPU were computed, let CTA0 broadcast the sum to over GPU
+            if (blockIdx.x == 0) {
+                volatile float * write_data =
+                    &((reinterpret_cast<float*>(params_pair_data))[data_offset]);
+
+                // write the data to memory region to be reflected to other GPU
+                asm volatile ("st.global.wt.v4.b32 [%0], {%1,%2,%3,%4};"
+                    :: "l"(write_data) , "f"(x[0]), "r"(magic), "f"(x[2]), "r"(magic));
+
+                asm volatile ("st.global.wt.v4.b32 [%0], {%1,%2,%3,%4};"
+                    :: "l"(write_data+4) , "f"(x[1]), "r"(magic), "f"(x[3]), "r"(magic));
             }
 
-            //now each CTA (on each GPU) reads the data written by CTA 0 of the other GPU
-            volatile float * read_data_ = &(((float*)params_my_data)[data_offset]);
-            volatile int32_t * read_flag = &(((int32_t*)((params_my_data)))[flags_offset]);
-
-            //check if other side has written
-#ifdef BNDEBUG
-            unsigned int safety=0;
-            
-            while ((read_flag[threadIdx.x] % 1000000) != (magic % 1000000) )
-            {
-                ++safety;
-                if (safety>99999) {
-                    printf("stuck waiting for my buddy, thread %d myvalue %d data offset %d flag offset %d read value %d\n", threadIdx.x, magic, 4*THREADS_PER_PIXEL+off*ELEMENTS_PER_LDG*THREADS_PER_PIXEL + ELEMENTS_PER_LDG*threadIdx.x, off*THREADS_PER_PIXEL, read_flag[threadIdx.x]);
-                    safety=0;
-                }
-            }
-#else
-            while ((read_flag[threadIdx.x] ) != (magic ) ) ;
-#endif
+            // now each CTA (on each GPU) reads the data written by CTA 0 of the other GPU
+            volatile float * read_data =
+                &((reinterpret_cast<float*>(params_my_data))[data_offset]);
 
             float other[4];
-                    asm volatile ("ld.global.cv.v4.f32 {%0, %1, %2, %3}, [%4];" 
-        : "=f"(other[0]), "=f"(other[1]), "=f"(other[2]), "=f"(other[3]) : "l"(read_data_));
+            uint32_t other_flag_a, other_flag_b;
+            do {
```

Comment:
So we removed the timeout from https://github.com/NVIDIA/apex/pull/360/files#diff-8a2b4400b4fad5e25aede628e56699c4L475
I'm totally OK with it if we think it's not necessary.

Reply:
that was dead debug code surrounded with `#ifdef BNDEBUG`

Reply:
Kinda unrelated question.
Is it necessary to have 1 magic flag per number? Would 1 flag per PTX instruction suffice? In this case it's not saving us anything, but I'm just curious.

Reply:
Previous version of the kernel used a single flag per 4 data elements and required a system fence to guarantee data arrival prior to flag update. By packing a flag+data into 64bit chunks here I am able to avoid the need for a fence. A different flag to data ratio given the current sizes would not allow a fence-less implementation.